pennylane/gradients/jvp.py

```diff
@@ -0,0 +1,287 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This module contains functions for computing the Jacobian vector product
+of tapes.
+"""
+import numpy as np
+
+import pennylane as qml
+from pennylane import math
+
+
+def _convert(jac, tangent):
+    """Utility to convert and cast the jacobian as tangent."""
+    if isinstance(jac, tuple):
+        jac_new = []
+        for j in jac:
+            j_ = math.convert_like(j, tangent)
+            j_ = math.cast(j_, tangent.dtype)
+            jac_new.append(j_)
+        jac = tuple(jac_new)
+    else:
+        jac = math.convert_like(jac, tangent)
+        jac = math.cast(jac, tangent.dtype)
+    return jac
+
+
+def compute_jvp_single(tangent, jac):
+    """Convenience function to compute the Jacobian vector product for a given
+    tangent vector and a Jacobian for a single measurement tape.
+
+    Args:
+        tangent (list, tensor_like): tangent vector
+        jac (tensor_like, tuple): Jacobian matrix
+    Returns:
+        tensor_like: the Jacobian vector product
+
+    **Examples**
+
+    1. For a single parameter and a single measurement without shape (e.g. expval, var):
+
+    .. code-block:: pycon
+        >>> tangent = np.array([1.0])
+        >>> jac = np.array(0.2)
+        >>> qml.gradients.compute_jvp_single(tangent, jac)
+        np.array(0.2)
+
+    2. For a single parameter and a single measurment with shape (e.g. probs):
+
+    .. code-block:: pycon
+        >>> tangent = np.array([2.0])
+        >>> jac = np.array([0.3, 0.3])
+        >>> qml.gradients.compute_jvp_single(tangent, jac)
+        np.array([0.6, 0.6])
+
+    3. For multiple parameters (in this case 2 parameters) and a single measurement without shape (e.g. expval, var):
+
+    .. code-block:: pycon
+        >>> tangent = np.array([1.0, 2.0])
+        >>> jac = tuple([np.array(0.1), np.array(0.2)])
+        >>> qml.gradients.compute_jvp_single(tangent, jac)
+        np.array(0.5)
+
+    4. For multiple parameters (in this case 2 parameters) and a single measurement with shape (e.g. probs):
+
+    .. code-block:: pycon
+        >>> tangent = np.array([1.0, 0.5])
+        >>> jac = tuple([np.array([0.1, 0.3]), np.array([0.2, 0.4])])
+        >>> qml.gradients.compute_jvp_single(tangent, jac)
+        np.array([0.2, 0.5])
+
+    """
+    if jac is None:
+        return None
+
+    tangent = qml.math.stack(tangent)
+    jac = _convert(jac, tangent)
+
+    # Single measurement with a single param
+    if not isinstance(jac, tuple):
+        tangent = math.reshape(tangent, (1,))
+        # Single measurement with no dimension e.g. expval
+        if jac.shape == ():
+            jac = math.reshape(jac, (1,))
+        # Single measurement with dimension e.g. probs
+        else:
+            jac = math.reshape(jac, (1, len(jac)))
```

Comment:
I ran into issues with `len` while working on the TensorFlow PR. My solution looked something like:
```suggestion
            jac = math.reshape(jac, (1, -1))
```